frontend/src/views/CalculateNutrients/CalculateNutrientsComponents/FertilizerModal.tsx

```diff
@@ -214,26 +344,152 @@ export default function FertilizerModal({
+      Object.entries(updates).forEach(([name, value]) => {
+        // eslint-disable-next-line prefer-const
+        let changes = structuredClone(updates);
 
-    if (name === 'fertilizerTypeId') {
-      setFilteredFertilizerOptions(
-        fertilizerOptions.filter(
-          (ele) =>
-            ele.dryliquid === fertilizerTypes.find((fertType) => fertType.id === value)?.dryliquid,
-        ),
-      );
-    }
+        if (['potassium', 'phosphorous', 'nitrogen'].includes(name)) {
+          setFormCustomFertilizer((prev) => ({ ...prev, ...changes }));
+        }
 
-    if (name === 'fertilizerId') {
-      // eslint-disable-next-line no-param-reassign
-      changes.name = filteredFertilizersOptions.find((f) => f.id === value)!.name;
-    }
+        if (name === 'fertilizerTypeId') {
+          setFilteredFertilizerOptions(
+            fertilizerOptions.filter(
+              (ele) =>
+                ele.dryliquid ===
+                fertilizerTypes.find((fertType) => fertType.id === value)?.dryliquid,
+            ),
+          );
 
-    setFormState((prev) => ({ ...prev, ...changes }));
-  };
+          if (value === 2) {
+            // For custom dry fertilizers
+            setFormCustomFertilizer((prev) => ({ ...prev, dryliquid: 'dry' }));
+          } else if (value === 4) {
+            // For custom liquid fertilizers
+            setFormCustomFertilizer((prev) => ({ ...prev, dryliquid: 'liquid' }));
+          } else {
+            // Reset for other values
+            setFormCustomFertilizer(EMPTY_CUSTOM_FERTILIZER);
+          }
+
+          // Reset other values on changes
+          changes = { ...structuredClone(EMPTY_FERTILIZER_FORM_DATA), fertilizerTypeId: value };
+        }
+
+        if (name === 'fertilizerId') {
+          changes.name = filteredFertilizersOptions.find((f) => f.id === value)!.name;
+          // Load liquid densities.
+          if (isLiquidFertilizer() && !isCustomFertilizer()) {
+            const densityValue = liqDensityFactors.find((ele) => ele.fertilizerid === value);
+
+            changes.density = densityValue.value;
+            changes.densityUnitId = densityValue.densityunitid;
+          }
+        }
+
+        if (name === 'densityUnitId') {
+          if (isLiquidFertilizer() && !isCustomFertilizer()) {
+            const densityValue = liqDensityFactors.find(
+              (ele) => ele.fertilizerid === formState.fertilizerId && ele.densityunitid === value,
+            );
+            changes.density = densityValue.value;
+            changes.densityUnitId = value;
+          }
+        }
+
+        setFormState((prev) => ({ ...prev, ...changes }));
+      });
+    },
+    [
+      fertilizerOptions,
+      fertilizerTypes,
+      filteredFertilizersOptions,
+      formState.fertilizerId,
+      isCustomFertilizer,
+      isLiquidFertilizer,
+      liqDensityFactors,
+    ],
+  );
+
+  const CustomFertilizerEntry = useMemo(() => {
+    if (isCustomFertilizer()) {
```

Comment:
Instead of all of these functions, since the functions don't take in any parameters, could we switch to `useMemo` booleans that depend on whatever stateful variables the function uses?

Reply:
Excellent idea. Switched to `useMemo` booleans.